csi: add nodeExpandSecret support for CSI client & add unit test

zhucan opened the pull request:
CSI spec 1.5 enhanced the spec to add optional secrets field to
NodeExpandVolumeRequest. This commit adds NodeExpandSecret to the
CSI PV source and also derive the expansion secret in csiclient to
send it out as part of the nodeexpand request.

#### What type of PR is this?
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #95367 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
This release add support for NodeExpandSecret for CSI driver client which enables the CSI drivers to make use of this secret while performing node expansion operation based on the user request. Previously there was no secret  provided as part of the nodeexpansion call, thus CSI drivers were not make use of the same while expanding the volume at node side. 
```
KEP reference：https://github.com/kubernetes/enhancements/pull/3173/

zhucan commented:
/test all

humblec commented:
This PR has the changes from https://github.com/kubernetes/kubernetes/pull/105958 and good for review. We have also tested the workflow of hostpath driver with the changes in external provisioner and its working . Please review with the target release of 1.23

humblec commented:
/assign @gnufied 
/assign @jsafrane 
/assign @msau42 
/assign @xing-yang

zhucan commented:
/test pull-kubernetes-verify

zhucan commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

xing-yang commented:
/milestone v1.23

jsafrane commented on pkg/apis/core/types.go:
```diff
@@ -1680,6 +1680,15 @@ type CSIPersistentVolumeSource struct {
+	// NodeExpandSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// NodeExpandVolume call.
+	// This is an alpha field and requires enabling ExpandCSIVolumes feature gate.
```
`ExpandCSIVolumes` is beta, not alpha.

zhucan replied:
done, both modify the ControllerExpandSecretRef

jsafrane commented:
You also need to drop `NodeExpandSecretRef` from a PV when the feature is disabled, like for `ControllerExpandSecretRef`: https://github.com/kubernetes/kubernetes/blob/76694983a1c50532206504e2b5e2b9767851354e/pkg/api/persistentvolume/util.go#L27-L33

Actually, please check where ExpandCSIVolumes feature is used, there may be more places like this.

zhucan commented:
> You also need to drop `NodeExpandSecretRef` from a PV when the feature is disabled, like for `ControllerExpandSecretRef`:
> 
> https://github.com/kubernetes/kubernetes/blob/76694983a1c50532206504e2b5e2b9767851354e/pkg/api/persistentvolume/util.go#L27-L33
> 
> Actually, please check where ExpandCSIVolumes feature is used, there may be more places like this.

done

jsafrane commented on pkg/api/persistentvolume/util.go:
```diff
@@ -28,6 +28,7 @@ func DropDisabledFields(pvSpec *api.PersistentVolumeSpec, oldPVSpec *api.Persist
 	if !utilfeature.DefaultFeatureGate.Enabled(features.ExpandCSIVolumes) && !hasExpansionSecrets(oldPVSpec) {
 		if pvSpec.CSI != nil {
 			pvSpec.CSI.ControllerExpandSecretRef = nil
+			pvSpec.CSI.NodeExpandSecretRef = nil
```
Please update the unit tests too.

zhucan replied:
added

jsafrane commented on pkg/api/persistentvolume/util.go:
```diff
@@ -28,6 +28,7 @@ func DropDisabledFields(pvSpec *api.PersistentVolumeSpec, oldPVSpec *api.Persist
 	if !utilfeature.DefaultFeatureGate.Enabled(features.ExpandCSIVolumes) && !hasExpansionSecrets(oldPVSpec) {
```
With this code, if there is ControllerExpandSecretRef  the old object and does not have `NodeExpandSecretRef`, the code does not clear `NodeExpandSecretRef` in the new object. Which is IMO wrong.

I think you should completely split `ControllerExpandSecretRef` and `NodeExpandSecretRef` handling, i.e. have `hasControllerExpansionSecrets()` and `hasNodeExpansionSecrets()` as separate functions and separate `if` here.

zhucan replied:
good suggestion

fedebongio commented:
/remove-sig api-machinery
/triage accepted

jsafrane commented on pkg/apis/core/types.go:
```diff
@@ -1675,11 +1675,20 @@ type CSIPersistentVolumeSource struct {
+	// NodeExpandSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// NodeExpandVolume call.
+	// This is an beta field and requires enabling ExpandCSIVolumes feature gate.
```
Heads up for an API reviewer: a new field goes straight to beta. From storage point of view, the field is needed and useful and it's basically a copy of well established `ControllerExpandSecretRef`.

msau42 replied:
We may need to add the field in alpha first. Otherwise a n-1 controller could drop the field on an update.

smarterclayton replied:
Yes, during an upgrade to "release with this" if someone set the field right away (because it was on by default) then it would get cleared if someone hit the not yet upgraded server.

Is there a justification for "straight to beta" recorded in the KEP discussion or other format?

smarterclayton replied:
I would prefer not to create the exception here to allow a field to go to beta ungated unless there is a decision by other api reviewers recorded already to do so.  It's unusual enough to give me pause, and michelle is right that it will have unusual implications for upgrades.

humblec replied:
Thanks all for the review and thoughts. As we are in a conclusion that this has to go through the feature gate starting with Alpha, this has to be tracked for next release.

jsafrane commented:
/lgtm
from storage perspective. It needs API review. Reading the [requirements](https://github.com/kubernetes/community/blob/master/sig-architecture/api-review-process.md#mechanics), you should update [CSI expansion KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/556-csi-volume-resizing/README.md#supporting-per-pvc-secret-refs), it still says that NodeExpand does not accept secrets.

xing-yang commented:
/assign @smarterclayton

smarterclayton commented:
Can you link the KEP in here (maybe I just missed it) so I can review that?

jsafrane commented:
@smarterclayton 
full KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/556-csi-volume-resizing/README.md
changes for the node secrets: https://github.com/kubernetes/enhancements/pull/3038

humblec commented:
@zhucan can you please rebase to master?

zhucan commented:
> @zhucan can you please rebase to master?

done

zhucan commented:
/retest

zhucan commented:
/test pull-kubernetes-e2e-gce-storage-slow

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1804,11 +1804,20 @@ type CSIPersistentVolumeSource struct {
 	// +optional
 	ControllerExpandSecretRef *SecretReference `json:"controllerExpandSecretRef,omitempty" protobuf:"bytes,9,opt,name=controllerExpandSecretRef"`
+
+	// NodeExpandSecretRef is a reference to the secret object containing
```
None of the field names in godoc in this struct are correct - they should all be `camelCase`, not `CamelCase`.  Can you open a follow up to make that change (not critical for this PR).

humblec replied:
sure, will track it with a follow up change. thanks!

xing-yang replied:
@zhucan Can you please address this comment?  Change "NodeExpandSecretRef" to "nodeExpandSecretRef".

zhucan commented:
/retest

zhucan commented:
/retest

zhucan commented:
/test pull-kubernetes-e2e-gce-storage-slow

humblec commented:
This PR followed the model of controllerExpandSecretRef and no separate feature gate was introduced. However based on the review and also based on the community  discussion, the decision is to put this under feature gate and take it from alpha from next release.  Thanks all ! will refresh the PR and continue on this .

reylejano commented:
/milestone clear
Release Managers are cutting 1.23.0-rc.0 and creating the `release-1.23` branch

dims commented:
Is this PR still needed, please rebase if so (or we can close it?)

humblec commented:
> Is this PR still needed, please rebase if so (or we can close it?)

we will be revamping this PR for a KEP @dims , can we take a decision to close or rebase in a few days?

humblec commented:
KEP #3107 track the enhancement requested here as `alpha` feature while `CSINodeExpandSecret` feature flag is enabled .

This PR will be updated soon to enable the feature flag as discussed in the KEP. 

Cc @zhucan

zhucan commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

humblec commented:
@gnufied @msau42 @xing-yang @jsafrane this is an updated version of the PR with feature flag in place, could you please take a look at this so that,  API review can be requested accordingly ?

zhucan commented:
/test pull-kubernetes-integration

xing-yang commented:
@zhucan Can you resolve the merge conflict?

zhucan commented:
> @zhucan Can you resolve the merge conflict?

ok

xing-yang commented on pkg/api/v1/persistentvolume/util_test.go:
```diff
@@ -149,6 +149,13 @@ func TestPVSecrets(t *testing.T) {
+			ClaimRef: &corev1.ObjectReference{Namespace: "claimrefns", Name: "claimrefname"},
+			PersistentVolumeSource: corev1.PersistentVolumeSource{
+				CSI: &corev1.CSIPersistentVolumeSource{
+					ControllerExpandSecretRef: &corev1.SecretReference{
```
Should be NodeExpandSecretRef.

xing-yang commented:
/lgtm

msau42 commented:
Can you put the generated files into a separate commit? That will make it easier to review

msau42 commented:
/label api-review

msau42 commented on pkg/api/persistentvolume/util.go:
```diff
@@ -25,14 +25,19 @@ import (
+				pvSpec.CSI.ControllerExpandSecretRef = nil
+			}
+			if !hasNodeExpansionSecrets(oldPVSpec) {
+				pvSpec.CSI.NodeExpandSecretRef = nil
```
This field check should be toggled by CSINodeExpandSecret feature gate

zhucan replied:
I had remove this change, I think it's not needed.

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -1716,11 +1716,20 @@ type CSIPersistentVolumeSource struct {
+	// NodeExpandSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// NodeExpandVolume call.
+	// This is an alpha field and requires enabling CSINodeExpand feature gate.
```
CSINodeExpandSecret feature gate

zhucan replied:
Your means update the "NodeExpandSecretRef" to  CSINodeExpandSecret? @msau42

msau42 replied:
The feature gate name in the comment is not correct. It should be "CSINodeExpandSecret"

msau42 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1801,11 +1801,20 @@ type CSIPersistentVolumeSource struct {
+	// nodeExpandSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// NodeExpandVolume call.
+	// This is an alpha field and requires enabling CSINodeExpand feature gate.
```
update CSINodeExpandSecret feature gate here too

msau42 commented on pkg/volume/csi/expander.go:
```diff
@@ -86,6 +86,15 @@ func (c *csiPlugin) nodeExpandWithClient(
 	}
+	nodeExpandSecrets := map[string]string{}
+	expandClient := c.host.GetKubeClient()
+	if csiSource.NodeExpandSecretRef != nil {
```
I think we need to check the feature gate here too

zhucan replied:
your means we should check the  CSINodeExpandSecret feature gate not  NodeExpandSecretRef?

msau42 replied:
In addition to checking `NodeExpandSecretRef`, also check the feature gate.

msau42 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1598,6 +1598,18 @@ func validateCSIPersistentVolumeSource(csi *core.CSIPersistentVolumeSource, fldP
 			allErrs = append(allErrs, ValidateDNS1123Label(csi.NodePublishSecretRef.Namespace, fldPath.Child("namespace"))...)
 		}
 	}
+	if csi.NodeExpandSecretRef != nil {
```
Add unit test

zhucan commented:
> /label api-review



> Can you put the generated files into a separate commit? That will make it easier to review

done

zhucan commented:
> Can you put the generated files into a separate commit? That will make it easier to review

done

zhucan commented:
/test pull-kubernetes-unit

zhucan commented:
/test pull-kubernetes-dependencies

zhucan commented:
/test pull-kubernetes-node-e2e-containerd

gnufied commented on pkg/api/v1/persistentvolume/util.go:
```diff
@@ -147,6 +147,11 @@ func VisitPVSecretNames(pv *corev1.PersistentVolume, visitor Visitor) bool {
 				return false
 			}
 		}
+		if source.CSI.NodeExpandSecretRef != nil {
```
Don't we need something in `DropDisabledFields` function of pv type, so as if feature is disabled and users try to set it, it gets dropped?

zhucan replied:
No needed,  if the feature-gates is disabled, we don't pass it to external-provisioner.

gnufied replied:
Not sure I understand. Regardless of what we do in external-provisioner, we need to drop the field if feature gate is not enabled. We carried similar logic for every other field while feature was in alpha.

gnufied replied:
See examples of other alpha fields being handled - https://github.com/kubernetes/kubernetes/commit/2c1b74376691cacf7b56904fe696df45753a8f99

gnufied commented on pkg/volume/csi/expander.go:
```diff
@@ -95,6 +106,7 @@ func (c *csiPlugin) nodeExpandWithClient(
 		accessMode:        api.ReadWriteOnce,
 		mountOptions:      pv.Spec.MountOptions,
+		secrets:           nodeExpandSecrets,
 	}
```
ideally we should have some tests for it.

zhucan commented:
/test pull-kubernetes-e2e-kind-ipv6

zhucan committed:
fix: add unit test

Signed-off-by: zhucan <zhucan.k8s@gmail.com>

gnufied commented:
@zhucan @humblec looks like the PR needs rebase?

msau42 reviewed (commented):
/lgtm
/approve

I have some minor nits but can be followed up later.

humblec commented:
> @zhucan @humblec looks like the PR needs rebase?

Sure, will rebase.. thanks.

humblec commented:
@msau42 the rebase took away `lgtm` .. could you please bring back..

humblec commented:
Cc @smarterclayton

zhucan commented:
/test pull-kubernetes-unit

zhucan commented on pkg/features/kube_features.go:
```diff
@@ -1004,5 +1011,6 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	// ...
-	HPAScaleToZero: {Default: false, PreRelease: featuregate.Alpha},
+	HPAScaleToZero:      {Default: false, PreRelease: featuregate.Alpha},
+	CSINodeExpandSecret: {Default: false, PreRelease: featuregate.Alpha},
```
done

zhucan commented:
/test pull-kubernetes-e2e-kind-ipv6

zhucan commented:
/test pull-kubernetes-unit

zhucan commented:
/test pull-kubernetes-e2e-gce-alpha-features

humblec commented:
/api-review

gnufied commented:
/lgtm

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1800,11 +1800,20 @@ type CSIPersistentVolumeSource struct {
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// NodeExpandVolume call.
+	// This is an alpha field and requires enabling CSINodeExpandSecret feature gate.
+	// This field is optional, and may be empty if no secret is required. If the
```
(I realize we're just copying the other fields, but it doesn't make sense to me as an API user what you mean by empty, so I'm trying to understand what it means so we can decide on a follow up)

msau42 replied:
probably should be "nil"

zhucan replied:
done

humblec replied:
@smarterclayton addressed this too.. ptal.. thanks !

zhucan commented:
/test pull-kubernetes-integration

zhucan commented:
/test pull-kubernetes-e2e-kind

zhucan commented:
/test pull-kubernetes-e2e-gce-alpha-features

humblec commented:
PR had to be rebased which took away lgtm... 

Cc @msau42 @smarterclayton @gnufied

xing-yang commented:
/lgtm

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1800,11 +1800,20 @@ type CSIPersistentVolumeSource struct {
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// NodeExpandVolume call.
+	// This is an alpha field and requires enabling CSINodeExpandSecret feature gate.
+	// This field is optional, and may be nil if no secret is required. If the
```
Sorry for this - but we wouldn't use `nil` because this apidoc is intended for JSON users (I should have clarified).  Now that I think you've answered the question of "the field is absent / omitted" I'd suggest this sentence (And all the others in this file, which can be done later) as:

> This field is optional, may be omitted if no secret is required.

humblec replied:
@smarterclayton sure,  will address this  in a follow up PR . Thanks!

zhucan replied:
done @humblec @smarterclayton

humblec replied:
Awesome. thanks @zhucan :+1:

humblec committed:
csi: add nodeExpandSecret support for CSI client

CSI spec 1.5 enhanced the spec to add optional secrets field to
NodeExpandVolumeRequest. This commit adds NodeExpandSecret to the
CSI PV source and also derive the expansion secret in csiclient to
send it out as part of the nodeexpand request.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>
Signed-off-by: zhucan <zhucan.k8s@gmail.com>

zhucan committed:
generated code and doc

Signed-off-by: zhucan <zhucan.k8s@gmail.com>

zhucan commented:
/test pull-kubernetes-e2e-kind

zhucan commented:
/test pull-kubernetes-integration

zhucan commented:
/test pull-kubernetes-e2e-kind

zhucan commented:
/test pull-kubernetes-integration

xing-yang commented:
/retest

zhucan commented:
/test pull-kubernetes-integration

xing-yang commented:
/lgtm

smarterclayton commented:
/approve

for api changes and field gate

humblec commented:
/test pull-kubernetes-unit

zhucan commented:
/test pull-kubernetes-unit

zhucan commented:
/test pull-kubernetes-e2e-gce-storage-slow

jsafrane commented:
/lgtm

xing-yang commented:
/milestone v1.25
